src/java/org/apache/cassandra/db/SystemKeyspaceMigrator41.java

```diff
@@ -159,10 +162,36 @@ static void migrateSSTableActivity()
                      })
         );
     }
+    
+    @VisibleForTesting
+    static void migrateCompactionHistory()
+    {
+        migrateTable(false,
+                     SystemKeyspace.COMPACTION_HISTORY,
+                     SystemKeyspace.COMPACTION_HISTORY,
+                     new String[]{ "id",
+                                   "bytes_in",
+                                   "bytes_out",
+                                   "columnfamily_name",
+                                   "compacted_at",
+                                   "keyspace_name",
+                                   "rows_merged",
+                                   "compaction_properties" },
+                     row -> Collections.singletonList(new Object[]{ row.getTimeUUID("id") ,
+                                                                    row.has("bytes_in") ? row.getLong("bytes_in") : null,
+                                                                    row.has("bytes_out") ? row.getLong("bytes_out") : null,
+                                                                    row.has("columnfamily_name") ? row.getString("columnfamily_name") : null,
+                                                                    row.has("compacted_at") ? row.getTimestamp("compacted_at") : null,
+                                                                    row.has("keyspace_name") ? row.getString("keyspace_name") : null,
+                                                                    row.has("rows_merged") ? row.getMap("rows_merged", Int32Type.instance, LongType.instance) : null,
+                                                                    CompactionHistoryProperty.getCompactionHistroyProperties(row.getMap("compaction_properties", UTF8Type.instance, UTF8Type.instance))
+                     })
+        );
+    }
 
     /**
      * Perform table migration by reading data from the old table, converting it, and adding to the new table.
-     *
+     * if oldName and newName are same, means refresh data in table 
      * @param truncateIfExists truncate the existing table if it exists before migration; if it is disabled
      *                         and the new table is not empty, no migration is performed
```

Comment:
`and the new table is not empty and newName is not equal to oldName, no migration is performed`

Reply:
btw @jacek-lewandowski this line in this method may produce null:

````
UntypedResultSet rows = QueryProcessor.executeInternal(query);
````

I suggest to either assert it is not null or if it is we log on warn level and return from this method.